File: js/Inverse_Distance_Weighted.js
<!-- Inverse Distance Weighted js -->
//输入[[x:0,y:0,v:0],[x:0,y:0,v:0],[x:0,y:0,v:0]]
function idwcomputer(datas,result){
   if(datas.lenght<3) return result;
   var m0=datas.length;
   var m1=result.length;
   
   //console.info(datas);
   
   //距离列表
   var r=[];
   
   for(var i=0;i<m1;i++){
       for(var j=0;j<m0;j++){
             var tmpDis = Math.sqrt(Math.pow(result[i].x - datas[j].x, 2) + Math.pow(result[i].y - datas[j].y, 2));
             r.push(tmpDis);
       }
   }
   
   //插值函数
 
   for (var i = 0; i < m1; i++)
   {
       //查找重复
       var ifFind = false;
       for (var j = m0 * i; j < m0 * i + m0; j++)
       {
           if (Math.abs(r[j]) < 0.0001)
           {
               result[i].v = datas[j - m0 * i].v;
               ifFind = true;
               break;
           }
       }
 
       if (ifFind) continue;
 
       var numerator = 0;
       var denominator = 0;
 
       for (var j = m0 * i; j < m0 * i + m0; j++)
       {
           numerator += datas[j - m0 * i].v / (r[j] * r[j]);
           denominator += 1 / (r[j] * r[j]);
       }
     
       result[i].v = numerator / denominator;
   }
   return result;

File: python/Inverse_Distance_Weighted.py
#Inverse Distance Weighted
import math
from math import radians, cos, sin, asin, sqrt
import copy
def get_h(lon,lat,lst,maxGeoDistanceMeter,sumCount):
    p0 = [lon, lat]
    sum0 = 0
    sum1 = 0
    temp = []
    # 遍历获取该点距离所有采样点的距离
    for point in lst:
        if lon == point["x"] and lat == point["y"]:
            return point["value"]
        #Di = distance(p0, point)
        Di=geodistance(p0[0],p0[1],point["x"],point["y"])
        # new出来一个对象，不然会改变原来lst的值
        ptn = copy.deepcopy(point)
        ptn["distance"]=Di
        temp.append(ptn)
 
    # 根据上面ptn.append（）的值由小到大排序
    temp1 = sorted(temp, key=lambda point: point["distance"])
    if(temp1[0]["distance"]>maxGeoDistanceMeter):
       return 0
    else:
        # 遍历排序的前15个点，根据公式求出sum0 and sum1
        for point in temp1[0:sumCount]:
            sum0 += point["value"] / point["distance"]
            sum1 += 1 / point["distance"]
    
        return sum0 / sum1
 
 
# 计算两点间的距离
def distance(p, pi):
    dis = (p[0] - pi[0]) * (p[0] - pi[0]) + (p[1] - pi[1]) * (p[1] - pi[1])
    m_result = math.sqrt(dis)
 
    return m_result
#计算两个经纬度之间的距离
def geodistance(lng1,lat1,lng2,lat2):
    lng1,lat1,lng2,lat2 = map(radians, [float(lng1), float(lat1), float(lng2), float(lat2)])
    dlon=lng2-lng1
    dlat=lat2-lat1
    a=sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
    distance=2*asin(sqrt(a))*6371*1000 # 地球平均半径，6371km
    distance=round(distance/1000,3)
    return distance*1000
